kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/LogReplay.java

```diff
@@ -18,94 +18,276 @@
-{
+/**
+ * Replays a history of actions, resolving them to produce the current state of the table. The
+ * protocol for resolution is as follows:
+ *  - The most recent {@link AddFile} and accompanying metadata for any `(path, dv id)` tuple wins.
+ *  - {@link RemoveFile} deletes a corresponding AddFile. A RemoveFile "corresponds" to the AddFile
+ *    that matches both the parquet file URI *and* the deletion vector's URI (if any).
+ *  - The most recent {@link Metadata} wins.
+ *  - The most recent {@link Protocol} version wins.
+ *  - For each `(path, dv id)` tuple, this class should always output only one {@link FileAction}
+ *    (either AddFile or RemoveFile)
+ *
+ * This class exposes only two public APIs
+ * - {@link #lazyLoadProtocolAndMetadata}: replay the log in reverse and return the first non-null
+ *                                         Protocol and Metadata
+ * - {@link #getAddFilesAsColumnarBatches}: return all active (not tombstoned) AddFiles as
+ *                                          {@link ColumnarBatch}s
+ */
+public class LogReplay implements Logging {
+
+    /** Read schema when searching for the latest Protocol and Metadata. */
+    public static final StructType PROTOCOL_METADATA_READ_SCHEMA = new StructType()
+        .add("protocol", Protocol.READ_SCHEMA)
+        .add("metaData", Metadata.READ_SCHEMA);
+
+    /** Read schema when searching for all the active AddFiles (need the RemoveFiles, too). */
+    public static final StructType ADD_REMOVE_READ_SCHEMA = new StructType()
+        .add("add", AddFile.READ_SCHEMA)
+        .add("remove", RemoveFile.READ_SCHEMA);
+
+    /** Data (result) schema of the remaining active AddFiles. */
+    public static final StructType ADD_ONLY_DATA_SCHEMA = new StructType()
+        .add("add", AddFile.READ_SCHEMA);
+
     private final Path dataPath;
     private final LogSegment logSegment;
-    private final CloseableIterable<Tuple2<Action, Boolean>> reverseActionsIterable;
+    private final TableClient tableClient;
+
+    /**
+     * An iterable generating iterators of ColumnarBatches of AddFiles and RemovedFiles, used to
+     * return the active AddFiles in the Snapshot.
+     */
+    private final CloseableIterable<Tuple2<FileDataReadResult, Boolean>>
+        addRemoveColumnarBatchIterable;
+
     private final Lazy<Tuple2<Protocol, Metadata>> protocolAndMetadata;
 
     public LogReplay(
-        Path logPath,
-        Path dataPath,
-        TableClient tableHelper,
-        LogSegment logSegment)
-    {
+            Path logPath,
+            Path dataPath,
+            TableClient tableClient,
+            LogSegment logSegment) {
         this.dataPath = dataPath;
         this.logSegment = logSegment;
+        this.tableClient = tableClient;
 
-        final Stream<FileStatus> allFiles = Stream.concat(
-            logSegment.checkpoints.stream(),
-            logSegment.deltas.stream());
+        final List<FileStatus> allFiles = Stream.concat(
```

Comment:
nit: rename `allFiles` to `deltaLogFiles`?

Reply:
this seems like a method that LogSegment should provide - getAllLogFiles(). better abstraction.